### DISCUSSION TITLE
iModel Text Overlay Embedding Font on Macbook

### DISCUSSION THREAD

**Discussion by PeterJBassett** [iTwin/itwinjs-core#2407] [Q&A]
Hello we have a bridge connector which includes some functionality to display a text overlay on an iModel. When this code is run on a non-windows machine, (eg. a Mac, or a Linux backend server) it throws an error trying to embed the font.

Please can you tell me if there is a solution?

Here is the code extract:

    let fontProps: FontProps = { id: 0, type: FontType.TrueType, name: "Arial" };
    try {
      fontProps = this.synchronizer.imodel.embedFont(fontProps); // throws Error
    } catch (error) {
      if ("win32" === Platform.platformName)
      console.log("Font embed failed");
    else
      console.log("Font embed failed. Not windows");
    }

.....

    for (const key in overlays) {
      if(!overlays.hasOwnProperty(key))
        continue;
      const overlay = overlays[key];
      const builder = new GeometryStreamBuilder();
      const textProps: TextStringProps ={
        text: overlay.text,
        font: fontProps.id,
        height: overlay.height * unitScale,
        widthFactor: 1
      }
      const textString = new TextString(textProps);
      builder.appendTextString(textString);
      if(!overlay.z)
        overlay.z = 0;
      let translation = Point3d.create(overlay.x, overlay.y, overlay.z)
      translation = this._bridgeParams.getGeoTransform().multiplyPoint3d(translation);
      translation.scaleInPlace(unitScale);
      const textElement: PhysicalElementProps = {
        model: physicalModelId,
        code: Code.createEmpty(),
        classFullName: PhysicalObject.classFullName,
        category: this.featureCategoryId,
        geom: builder.geometryStream,
        userLabel: overlay.text,
        placement: new Placement3d(translation, YawPitchRollAngles.createDegrees(0, 0, 0), Range3d.createNull())
      };
      this.synchronizer.imodel.elements.insertElement(textElement);
    }

**Comment by pmconne** [comment]
@jffmarker @abeesh is this a known issue?

**Reply by bbastings** [reply]
Yes, it's a known limitation. We need to provide a different mechanism for locating fonts that does not require embedding them.

**Reply by jffmarker** [reply]
Well, embedding may still ultimately be required, but allowing sources other than the Win32 API global font API (such as a hosted service) would also be beneficial. Regardless, from a native standpoint, if the system running the Connector has the file file(s) available, `DgnFontPersistence::File::FromTrueTypeFiles` could be used to create a font object, and you should be able to embed as normal from there... I'm not sure if this was exposed through JS... @bbastings?

**Reply by PeterJBassett** [reply]
Hello I am just writing to ask if there is any further update on this issue. 

It is still a problem for the FutureOn connector to iModel. The FutureOn connector is deployed using a linux docker image running in a kubernetes infrastructure.

Is there any javascript solution that will work in a linux container?

**Reply by pmconne** [reply]
@kabentley is currently working on this in #3080.

**Reply by kabentley** [reply]
@PeterJBassett, As @pmconne mentions, [This PR](https://github.com/iTwin/itwinjs-core/pull/3080) addresses this issue in 2 ways:

1) allowing fonts to be loaded from "workspaces files" that are external to iModels. Ultimately this should be the preferred way to ensure fonts are available when needed. Embedding fonts _within_ an iModel should only be needed for "freezing" an iModel, much like embedding fonts in a PDF file.
2) expanding the "embed" api so that it is possible to embed fonts on non-windows platforms. To do that on Linux, you'll need to supply the filename of the font (e.g. .ttf, ttc., .shx, etc.).

I expect to complete the PR this week.

HTH,
Keith
